File: websocket-api/csharp/UpdateOrder.cs
using System;
using System.Net.WebSockets;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

class UpdateOrder
{
    static async Task Main(string[] args)
    {
        DotNetEnv.Env.Load();
        
        string apiKey = Environment.GetEnvironmentVariable("API_KEY");
        string secretKey = Environment.GetEnvironmentVariable("SECRET_KEY");
        
        if (string.IsNullOrEmpty(apiKey) || string.IsNullOrEmpty(secretKey))
        {
            throw new Exception("Required environment variables (API_KEY, SECRET_KEY) are missing.");
        }
        
        using (var ws = new ClientWebSocket())
        {
            await ws.ConnectAsync(new Uri("wss://ws-api.ripio.com"), CancellationToken.None);
            Console.WriteLine("Connected to WebSocket API");
            
            long timestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            var bodyParams = new
            {
                id = "7155ED34-9EC4-4733-8B32-1E4319CB662F", // Replace with actual order ID
                price = "350000",
                amount = "0.02"
            };
            
            string bodyJson = JsonSerializer.Serialize(bodyParams);
            string signature = GenerateSignature(timestamp, bodyJson, secretKey);
            
            var request = new
            {
                id = "req-update-001",
                method = "order.update",
                @params = new
                {
                    id = bodyParams.id,
                    price = bodyParams.price,
                    amount = bodyParams.amount,
                    apiToken = apiKey,
                    timestamp = timestamp,
                    signature = signature
                }
            };
            
            string requestJson = JsonSerializer.Serialize(request);
            Console.WriteLine($"\nSending request: {requestJson}");
            
            byte[] buffer = Encoding.UTF8.GetBytes(requestJson);
            await ws.SendAsync(new ArraySegment<byte>(buffer), WebSocketMessageType.Text, true, CancellationToken.None);
            
            var receiveBuffer = new byte[4096];
            var result = await ws.ReceiveAsync(new ArraySegment<byte>(receiveBuffer), CancellationToken.None);
            string response = Encoding.UTF8.GetString(receiveBuffer, 0, result.Count);
            Console.WriteLine($"\nReceived response: {response}");
            
            await ws.CloseAsync(WebSocketCloseStatus.NormalClosure, "Done", CancellationToken.None);
        }
    }
    
    static string GenerateSignature(long timestamp, string bodyJson, string secretKey)
    {
        string message = timestamp.ToString() + bodyJson;
        using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(secretKey)))
        {
            byte[] hash = hmac.ComputeHash(Encoding.UTF8.GetBytes(message));
            return Convert.ToBase64String(hash);
        }
    }
}

File: websocket-api/python/update_order.py
import asyncio
import json
import hmac
import hashlib
import base64
import time
import os
import websockets
from dotenv import load_dotenv

load_dotenv()

if os.getenv('API_KEY') is None or os.getenv('SECRET_KEY') is None:
    raise ValueError("Required environment variables (API_KEY, SECRET_KEY) are missing.")

api_key = os.getenv('API_KEY')
secret_key = os.getenv('SECRET_KEY')

# WebSocket endpoint
WS_URL = 'wss://ws-api.ripio.com'

def generate_signature(timestamp, body_params):
    """Generate HMAC SHA256 signature for WebSocket API"""
    message = str(timestamp) + json.dumps(body_params, separators=(',', ':'))
    signature = hmac.new(
        secret_key.encode(),
        message.encode(),
        hashlib.sha256
    ).digest()
    return base64.b64encode(signature).decode()

async def update_order():
    async with websockets.connect(WS_URL) as websocket:
        print('Connected to WebSocket API')

        # Prepare update parameters (business params only for signature)
        # Replace with an actual order ID from your account
        timestamp = int(time.time() * 1000)
        body_params = {
            'id': '7155ED34-9EC4-4733-8B32-1E4319CB662F',  # Replace with actual order ID
            'price': '350000',
            'amount': '0.02'
        }

        # Generate signature
        signature = generate_signature(timestamp, body_params)

        # Create complete request with authentication
        request = {
            'id': 'req-update-001',
            'method': 'order.update',
            'params': {
                **body_params,
                'apiToken': api_key,
                'timestamp': timestamp,
                'signature': signature
            }
        }

        print('\nSending update order request:', json.dumps(request, indent=2))
        await websocket.send(json.dumps(request))

        # Receive response
        response = await websocket.recv()
        response_data = json.loads(response)
        print('\nReceived response:', json.dumps(response_data, indent=2))

if __name__ == '__main__':
    asyncio.run(update_order())
